### ISSUE TITLE
Osquery doesn't re-register in Audit Framework/Subsystem after restart

### ISSUE THREAD

**Issue by macEar** [osquery/osquery#6179]
# Bug report

### What operating system and version are you using?

```json
"columns":{"build":"","codename":"","major":"7","minor":"7","name":"Red Hat Enterprise Linux Server","patch":"0","platform":"rhel","platform_like":"rhel","version":"Red Hat Enterprise Linux Server release 7.7 (Maipo)"}​
```
### What version of osquery are you using?

```json
{"build_distro":"centos7","build_platform":"1","config_hash":"ecc935236418e2d89eac8e2388215516faf99dcd","config_valid":"1","extensions":"active","instance_id":"8e99e779-f77a-4893-922f-05e2a3aabdfd","pid":"70739","start_time":"1579027594","uuid":"44D06BF1-66B3-6343-A96F-48A84715064B","version":"4.0.2","watcher":"113773"}​
```

### What steps did you take to reproduce the issue?

I've got some flags in Kolide Fleet set (not all flags):

```
...
disable_watchdog: true
watchdog_memory_limit: 550
watchdog_utilization_limit: 720
watchdog_delay: 480​
disable_audit: false
audit_allow_config: true
audit_allow_process_events: true
audit_allow_fim_events: false
audit_allow_sockets: false
audit_persist: true​
...
```

After osqueryd has been started I checked audit rules by:

```
[root@... log]# auditctl -l
-a always,exit -S execve
-a always,exit -S execveat
```

It says process auditing is on, as I set it. After that I checked PID of the osqueryd process:

```
[root@... rules.d]# ps -ef | grep osquery
root       8999  39833  0 14:52 pts/3    00:00:00 grep --color=auto osquery
root      71218      1  0 14:36 ?        00:00:00 /usr/bin/osqueryd --flagfile /etc/osquery/osquery.flags --config_path /etc/osquery/osquery.conf
root      71261  71218  4 14:36 ?        00:00:45 /usr/bin/osqueryd
```

PID equals to 71261, OK. Now issue the following command:

```
[root@... rules.d]# auditctl -s
enabled 1
failure 0
pid 71261
rate_limit 0
backlog_limit 4096
lost 116302
backlog 0
loginuid_immutable 0 unlocked
```

That means everything is ok as PIDs are the same.

Restart osqueryd by:

```
osqueryctl restart
```

And its PID is changed, but in output of auditctl -s command the old one remained. The few minutes after PID in auditctl switched to 0.

```
[root@... log]# auditctl -s
enabled 1
failure 0
pid 0
rate_limit 0
backlog_limit 4096
lost 24433555
backlog 0
```

Due to auditd option -f 1, which is default, I notice lots of auditd messages in logs, it's overwhelmed:

```
[11021106.669324] audit: type=1327 audit(1578909143.893:180613784): proctitle=6F7261636C656F646200284C4F43414C3D4E4F29
[11021106.671836] audit: type=1300 audit(1578909143.896:180613785): arch=c000003e syscall=59 success=yes exit=0 a0=1696400 a1=162ca90 a2=1679b30 a3=7ffc856dd760 items=2 ppid=1 pid=102292 auid=1101 uid=1000 gid=1001 euid=1000 suid=1000 fsuid=1000 egid=1001 sgid=1001 fsgid=1001 tty=(none) ses=1486808 comm="oracle" exe="/u/app/oracle/base/product/12_inv/bin/oracle" subj=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023 key=(null)
[11021106.671838] audit: type=1309 audit(1578909143.896:180613785): argc=2 a0="oracleodb" a1="(LOCAL=NO)"
[11021106.671839] audit: type=1307 audit(1578909143.896:180613785): cwd="/u/app/oracle"
[11021106.671841] audit: type=1302 audit(1578909143.896:180613785): item=0 name="/u/app/oracle/base/product/12_inv/bin/oracle" inode=9990070 dev=fc:10 mode=0106751 ouid=1000 ogid=1001 rdev=00:00 obj=unconfined_u:object_r:default_t:s0 nametype=NORMAL cap_fp=0000000000000000 cap_fi=0000000000000000 cap_fe=0 cap_fver=0
```

### What did you expect to see?

Process auditing works fine and there is no problems with osquery registering in Linux Audit Framework/Subsystem. Logs aren't overwhelmed by auditd messages.

### What did you see instead?

Any time osqueryd restarts, in auditctl -s pid is 0 and lots of auditd messages are in logs. 
It's worth noticing that auditd is disabled and off. 
For example, osuqeyrd can restart with the following log message (or with CPU limit exceeded):

```
osqueryd worker (25673) stopping: Memory limits exceeded: 372044000​
```

After such restarts osqueryd PID is changed and the above-mentioned problem occurs.

And another question: does osquery drop those rules which it made enabled after it quits or restarts? Seems that the answer is no.

**Comment by macEar** [comment]
Also did the following steps:

```
sudo auditctl -D
sudo auditctl -a always,exit -S execve
sudo auditctl -a always,exit -S execveat
sudo auditctl -l
sudo rm /var/log/audit/*.log
sudo service auditd start; sleep 30; sudo service auditd stop
auditctl -D

for SYSCALL in execve execveat
  do
    echo -e "\nReport for $SYSCALL syscall"
    ausearch --syscall $SYSCALL --line-buffered --raw | aureport --executable --summary | head -15
done
```

В результате:

```
Report for execve syscall

Executable Summary Report
=================================
total  file
=================================
1  /usr/bin/bash
1  /usr/bin/mktemp
1  /usr/bin/gawk
1  /usr/bin/ls
1  /usr/bin/grep
1  /usr/bin/cat
1  /usr/bin/cmp
1  /usr/bin/rm
1  /usr/sbin/auditctl

Report for execveat syscall

Executable Summary Report
=================================
total  file
=================================
<no events of interest were found>
```

**Comment by theopolis** [comment]
You mention that the `auditd` service is off but is the `auditd` process still running `ps aux | grep auditd` should show nothing or just the kernel thread `[auditd]`?

I ask this because it might be the case that when `osqueryd` restarts it tries to regain control of the audit netlink socket and fails because the `auditd` process already has control.

**Comment by macEar** [comment]
Checked that, found only ```[kauditd]``` running in kernel space.

Also tried to reproduce this issue on the __standby server__, and have the following result:

* Check audit rules loaded:

```
[root@... log]# auditctl -l
-a always,exit -S execve
-a always,exit -S execveat
```

* Check audit configuration parameters:

```
[root@... log]# auditctl -s
enabled 1
failure 0
pid 11569
rate_limit 0
backlog_limit 4096
lost 0
backlog 0
loginuid_immutable 0 unlocked
```

* Check osqueryd PID:

```
[root@... log]# ps -ef | grep osquery
root      2651     1  0  2019 ?        00:16:20 /usr/bin/osqueryd --flagfile /etc/osquery/osquery.flags --config_path /etc/osquery/osquery.conf
root      3497  2139  0 09:20 pts/4    00:00:00 grep --color=auto osquery
root     11569  2651  5 08:30 ?        00:02:54 /usr/bin/osqueryd
```

* Check if any limit exceeded messages appeared in logs:

```
[root@... log]# grep "osqueryd worker" /var/log/messages
[root@... log]#
```

* Restart osqueryd, it changed its PID and audit framework registered new one right away:

```
[root@... log]# systemctl restart osqueryd
[root@... log]# ps -ef | grep osquery
root      8033     1  0 09:27 ?        00:00:00 /usr/bin/osqueryd --flagfile /etc/osquery/osquery.flags --config_path /etc/osquery/osquery.conf
root      8035  8033  3 09:27 ?        00:00:00 /usr/bin/osqueryd
root      8094  2139  0 09:27 pts/4    00:00:00 grep --color=auto osquery
[root@... log]# auditctl -s
enabled 1
failure 0
pid 8035
rate_limit 0
backlog_limit 4096
lost 0
backlog 0
loginuid_immutable 0 unlocked
```

After some playing around we noticed that:

* If you restart osqueryd manually (```osqueryctl restart``` or in our case ```systemctl restart osqueryd```), everything is fine as Audit Framework registers new PID immediately.

* If osqueryd is restarted by the watcher process (because of memory or CPU limit exceeded even regardless of the fact that we turned off ```watchdog```), pid isn't changed in Audit Subsystem and it constitutes the problem. After it was restarted by the watcher osqueryd got new pid:

```
root      1407     1  0 Jan16 ?        00:00:30 /usr/bin/osqueryd --flagfile=/etc/osquery/osquery.flags --pidfile=/var/run/osqueryd.pidfile --daemonize=true
root      1409  1407  0 Jan16 ?        00:01:58 /usr/bin/osqueryd​
```

but in ```auditctl -s``` pid wasn't changed and after some time it turned to 0.

* Checked on the __standby server__ for all logs where there are either ```exceeded``` or ```started``` word in them to understand when after restart we got limit exceeded messages.

```
5 | 1 | 2020-01-17 07:01:12 | 2020-01-17 07:01:12 | ... | ... |   | osqueryd started [version=4.0.2] |   | osqueryd | - | - |  
5 | 1 | 2020-01-17 06:27:05 | 2020-01-17 06:27:05 | ... | ... |   | osqueryd started [version=4.0.2] |   | osqueryd | - | - |  
5 | 1 | 2020-01-17 05:30:06 | 2020-01-17 05:30:06 | ... | ... |   | osqueryd worker (13523) stopping: Memory limits exceeded: 387864000 |   | osqueryd | - | - |  
6 | 3 | 2020-01-17 05:30:06 | 2020-01-17 05:30:06 | ... | ... |   | W0117 08:30:06.931551 3028 watcher.cpp:331] osqueryd worker (13523) stopping: Memory limits exceeded: 387864000 |   | osqueryd | - | - |  
5 | 1 | 2020-01-17 04:30:08 | 2020-01-17 04:30:08 | ... | ... |   | osqueryd worker (27318) stopping: Memory limits exceeded: 391216000 |   | osqueryd | - | - |  
6 | 3 | 2020-01-17 04:30:08 | 2020-01-17 04:30:08 | ... | ... |   | W0117 07:30:08.291014 3028 watcher.cpp:331] osqueryd worker (27318) stopping: Memory limits exceeded: 391216000 |   | osqueryd | - | - |  
5 | 1 | 2020-01-16 21:30:05 | 2020-01-16 21:30:05 | ... | ... |   | osqueryd worker (23834) stopping: Memory limits exceeded: 395572000 |   | osqueryd | - | - |  
6 | 3 | 2020-01-16 21:30:05 | 2020-01-16 21:30:05 | ... | ... |   | W0117 00:30:05.189069 3028 watcher.cpp:331] osqueryd worker (23834) stopping: Memory limits exceeded: 395572000 |   | osqueryd | - | - |  
5 | 1 | 2020-01-16 18:15:09 | 2020-01-16 18:15:09 | ... | ... |   | osqueryd worker (33402) stopping: Memory limits exceeded: 377832000 |   | osqueryd | - | - |  
6 | 3 | 2020-01-16 18:15:09 | 2020-01-16 18:15:09 | ... | ... |   | W0116 21:15:09.749413 3028 watcher.cpp:331] osqueryd worker (33402) stopping: Memory limits exceeded: 377832000 |   | osqueryd | - | - |  
5 | 1 | 2020-01-16 11:10:06 | 2020-01-16 11:10:06 | ... | ... |   | osqueryd worker (3152) stopping: Memory limits exceeded: 376364000 |   | osqueryd | - | - |  
6 | 3 | 2020-01-16 11:10:06 | 2020-01-16 11:10:06 | ... | ... |   | W0116 14:10:06.592367 3028 watcher.cpp:331] osqueryd worker (3152) stopping: Memory limits exceeded: 376364000 |   | osqueryd | - | - |  
5 | 1 | 2020-01-16 06:15:06 | 2020-01-16 06:15:06 | ... | ... |   | osqueryd worker (40676) stopping: Memory limits exceeded: 372780000 |   | osqueryd | - | - |  
6 | 3 | 2020-01-16 06:15:06 | 2020-01-16 06:15:06 | ... | ... |   | W0116 09:15:06.172955 3028 watcher.cpp:331] osqueryd worker (40676) stopping: Memory limits exceeded: 372780000 |   | osqueryd | - | - |  
5 | 1 | 2020-01-16 00:55:04 | 2020-01-16 00:55:04 | ... | ... |   | osqueryd worker (22437) stopping: Memory limits exceeded: 369464000 |   | osqueryd | - | - |  
6 | 3 | 2020-01-16 00:55:04 | 2020-01-16 00:55:04 | ... | ... |   | W0116 03:55:04.962019 3028 watcher.cpp:331] osqueryd worker (22437) stopping: Memory limits exceeded: 369464000 |   | osqueryd | - | - |  
5 | 1 | 2020-01-15 22:20:10 | 2020-01-15 22:20:10 | ... | ... |   | osqueryd worker (20158) stopping: Memory limits exceeded: 378896000 |   | osqueryd | - | - |  
6 | 3 | 2020-01-15 22:20:10 | 2020-01-15 22:20:10 | ... | ... |   | W0116 01:20:10.364524 3028 watcher.cpp:331] osqueryd worker (20158) stopping: Memory limits exceeded: 378896000 |   | osqueryd | - | - |  
5 | 1 | 2020-01-15 14:00:05 | 2020-01-15 14:00:05 | ... | ... |   | osqueryd worker (27389) stopping: Memory limits exceeded: 367012000 |   | osqueryd | - | - |  
6 | 3 | 2020-01-15 14:00:05 | 2020-01-15 14:00:05 | ... | ... |   | W0115 17:00:05.269383 3028 watcher.cpp:331] osqueryd worker (27389) stopping: Memory limits exceeded: 367012000 |   | osqueryd | - | -​
```

Also did the same for the __primary server__:

```
6 | 3 | 2020-01-17 09:36:46 | 2020-01-17 09:36:46 | ... | ... |   | W0117 12:36:46.084977 60678 watcher.cpp:331] osqueryd worker (60679) stopping: Memory limits exceeded: 379908000 |   | osqueryd | - | - |  
5 | 1 | 2020-01-17 09:36:46 | 2020-01-17 09:36:46 | ... | ... |   | osqueryd worker (60679) stopping: Memory limits exceeded: 379908000 |   | osqueryd | - | - |  
5 | 1 | 2020-01-17 09:00:41 | 2020-01-17 09:00:41 | ... | ... |   | osqueryd started [version=4.0.2] |   | osqueryd | - | - |  
5 | 1 | 2020-01-17 08:26:43 | 2020-01-17 08:26:43 | ... | ... |   | osqueryd started [version=4.0.2] |   | osqueryd | - | - |  
5 | 1 | 2020-01-17 08:04:23 | 2020-01-17 08:04:23 | ... | ... |   | osqueryd worker (18781) stopping: Memory limits exceeded: 374460000 |   | osqueryd | - | - |  
6 | 3 | 2020-01-17 08:04:23 | 2020-01-17 08:04:23 | ... | ... |   | W0117 11:04:23.983716 18780 watcher.cpp:331] osqueryd worker (18781) stopping: Memory limits exceeded: 374460000 |   | osqueryd | - | - |  
5 | 1 | 2020-01-17 07:33:52 | 2020-01-17 07:33:52 | ... | ... |   | osqueryd started [version=4.0.2] |   | osqueryd | - | - |  
5 | 1 | 2020-01-17 06:58:13 | 2020-01-17 06:58:13 | ... | ... |   | osqueryd started [version=4.0.2] |   | osqueryd | - | - |  
5 | 1 | 2020-01-17 06:55:32 | 2020-01-17 06:55:32 | ... | ... |   | osqueryd started [version=4.0.2] |   | osqueryd | - | - |  
5 | 1 | 2020-01-17 06:53:43 | 2020-01-17 06:53:43 | ... | ... |   | osqueryd started [version=4.0.2] |   | osqueryd | - | - |  
5 | 1 | 2020-01-17 06:32:06 | 2020-01-17 06:32:06 | ... | ... |   | osqueryd started [version=4.0.2] |   | osqueryd | - | - |  
5 | 1 | 2020-01-16 14:14:47 | 2020-01-16 14:14:47 | ... | ... |   | osqueryd worker (57139) stopping: Memory limits exceeded: 380164000 |   | osqueryd | - | - |  
6 | 3 | 2020-01-16 14:14:47 | 2020-01-16 14:14:47 | ... | ... |   | W0116 17:14:47.384290 57138 watcher.cpp:331] osqueryd worker (57139) stopping: Memory limits exceeded: 380164000 |   | osqueryd | - | - |  
5 | 1 | 2020-01-16 13:26:44 | 2020-01-16 13:26:44 | ... | ... |   | osqueryd started [version=4.0.2] |   | osqueryd | - | - |  
5 | 1 | 2020-01-16 07:49:35 | 2020-01-16 07:49:35 | ... | ... |   | osqueryd worker (77079) stopping: Memory limits exceeded: 367672000 |   | osqueryd | - | - |  
6 | 3 | 2020-01-16 07:49:35 | 2020-01-16 07:49:35 | ... | ... |   | W0116 10:49:35.584861 77069 watcher.cpp:331] osqueryd worker (77079) stopping: Memory limits exceeded: 367672000 |   | osqueryd | - | - |  
5 | 1 | 2020-01-16 07:13:31 | 2020-01-16 07:13:31 | ... | ... |   | osqueryd started [version=4.0.2] |   | osqueryd | - | - |  
5 | 1 | 2020-01-15 13:46:52 | 2020-01-15 13:46:52 | ... | ... |   | osqueryd worker (25673) stopping: Memory limits exceeded: 372044000 |   | osqueryd | - | - |  
6 | 3 | 2020-01-15 13:46:52 | 2020-01-15 13:46:52 | ... | ... |   | W0115 16:46:52.585137 25670 watcher.cpp:331] osqueryd worker (25673) stopping: Memory limits exceeded: 372044000 |   | osqueryd | - | - |  
5 | 1 | 2020-01-15 13:10:49 | 2020-01-15 13:10:49 | ... | ... |   | osqueryd started [version=4.0.2] |   | osqueryd | - | -​
```

As a result, it seems that there is a problem with re-registering in Audit Framework only after __watcher__ causes restart of osqueryd because of memory limit exceeded. But restarting manually works as it is supposed to work.

**Comment by macEar** [comment]
Added logs from those servers for last 3 days excluding query results.

[standby server logs.txt](https://github.com/osquery/osquery/files/4080132/standby.server.logs.txt)
[primary server logs.txt](https://github.com/osquery/osquery/files/4080133/primary.server.logs.txt)

Found in them some additional error messages.

**Comment by macEar** [comment]
@theopolis 
how can I check whether the audit socket is busy by another process?

**Comment by macEar** [comment]
The problem solved for now. 

osqueryd for some reason utilizied a lot of CPU which leads to osqueryd being constantly restarted by watchdog (or wathcer). All those before-mentioned problems are consequences of it.

I checked on /var/osquery/osquery.db/* and the size of all files in this directory was 2 Gb. Then I stopped osqueryd, deleted all the files and started it again. Eventually, osqueryd ceased to utilize a lot of CPU and right now working as it is supposed to.
